spark-knn-core/src/main/scala/org/apache/spark/ml/knn/KNN.scala

```diff
@@ -227,24 +248,36 @@ class KNNModel private[ml](
   /** @group setParam */
   def setBufferSize(value: Double): this.type = set(bufferSize, value)
 
-  //TODO: All these can benefit from DataSet API in Spark 1.6
-  override def transform(dataset: DataFrame): DataFrame = {
-    val merged = transform(dataset, topTree, subTrees)
+  //TODO: All these can benefit from DataSet API
+  override def transform(dataset: Dataset[_]): DataFrame = {
+    val merged: RDD[(Long, Array[(Row,Double)])] = transform(dataset, topTree, subTrees)
+
+    val withDistance = $(distanceCol).nonEmpty
 
     dataset.sqlContext.createDataFrame(
-      dataset.rdd.zipWithIndex().map { case (row, i) => (i, row) }
-        .leftOuterJoin(merged)
+      dataset.toDF().rdd.zipWithIndex().map { case (row, i) => (i, row) }
+        .leftOuterJoin(merged.asInstanceOf[RDD[(Long,Array[(Row, Double)])]])
```

Comment:
Is `.asInstanceOf[RDD[(Long,Array[(Row, Double)])]]` necessary since the val already has explicit type above?